### PR TITLE
Adding new metric type to track in-flight queries via the remote read API endpoint.

### LINKED ISSUE #4661: Metric for number of remote_read queries in-flight
We have a metric that shows the number in-flight queries on the promql engine: `prometheus_engine_queries`, with the introduction of the concurrency limits #4656, a similar metric should be added for in-flight remote-read queries.

### PR DESCRIPTION
**Pull request by tariq1890**
Fixes #4661 @gouthamve

### CODE REVIEW

File: web/api/v1/api.go [hunk 1]
```diff
@@ -28,9 +28,11 @@ import (
 	"time"
 	"unsafe"
 
+	"github.com/prometheus/client_golang/prometheus"
+
 	"github.com/go-kit/kit/log"
 	"github.com/go-kit/kit/log/level"
-	jsoniter "github.com/json-iterator/go"
+	"github.com/json-iterator/go"
```
Comment (simonpasquier) [inline web/api/v1/api.go]: This change isn't needed, I suspect that the alias is there to make obvious that `github.com/json-iterator/go` == `jsoniter`

**Comment by tariq1890** [comment]
@simonpasquier Thanks for reviewing. I have addressed your comments :)

File: web/api/v1/api.go [hunk 2]
```diff
@@ -28,6 +28,8 @@ import (
 	"time"
 	"unsafe"
 
+	"github.com/prometheus/client_golang/prometheus"
```
Comment (simonpasquier) [inline web/api/v1/api.go]: The package import should go with the other 3rd party packages just below.

File: web/api/v1/api.go [hunk 3]
```diff
@@ -781,10 +795,12 @@ func (api *API) remoteRead(w http.ResponseWriter, r *http.Request) {
 		}
 
 		querier, err := api.Queryable.Querier(r.Context(), from, through)
+		remoteReadQueries.Inc()
```
Comment (simonpasquier) [inline web/api/v1/api.go]: This should be done at the beginning of the function. By the way, the metric's update could be delegated to `api.remoteReadGate` as we want to increase the gauge at the beginning of `api.remoteReadGate.Start()` and decrease it when `api.remoteReadGate.End()` returns.

@gouthamve WDYT?

**Comment by cstyan** [comment]
I think including the remote URL as a label in the metrics would be useful. This could be done in a similar way to how we include the queue, which contains some info about that queues remote endpoint: https://github.com/prometheus/prometheus/blob/master/storage/remote/queue_manager.go#L189

File: web/api/v1/api.go (hunk 3 above)
Comment (cstyan) [reply web/api/v1/api.go]: this makes sense to me

**Comment by tariq1890** [comment]
Done. I've also implemented Gauge vector as per @cstyan's suggestions.

File: web/api/v1/api.go [hunk 4]
```diff
@@ -762,7 +778,10 @@ func (api *API) serveFlags(r *http.Request) (interface{}, *apiError, func()) {
 
 func (api *API) remoteRead(w http.ResponseWriter, r *http.Request) {
 	api.remoteReadGate.Start(r.Context())
+	remoteReadQueries.WithLabelValues(r.RequestURI).Inc()
```
Comment (gouthamve) [inline web/api/v1/api.go]: What does a typical `r.RequestURI` look like? Isn't it going to be same every time?

If it's going to have the params or selectors too, then we'll have the infinite-cardinality issue.

File: web/api/v1/api.go [hunk 5]
```diff
@@ -78,6 +85,15 @@ var corsHeaders = map[string]string{
...
 }
 
+var remoteReadQueries = prometheus.NewGaugeVec(
+	prometheus.GaugeOpts{
+		Namespace: namespace,
+		Subsystem: subsystem,
+		Name:      "remote_read_queries",
+		Help:      "The current number of remote read queries being executed or waiting.",
+	},
+	[]string{urlLabel})
```
Comment (gouthamve) [inline web/api/v1/api.go]: No need to make a new variable. Just out `[]string{"url"}` here.

Comment (tariq1890) [reply web/api/v1/api.go]: alright

File: web/api/v1/api.go (hunk 4 above)
Comment (tariq1890) [reply web/api/v1/api.go]: Yes. you are definitely correct. What would you recommend as a label ? In my previous commit I just added this metric without a label. Let me know if we are better off not having a label for this

Comment (tariq1890) [reply web/api/v1/api.go]: Done @gouthamve

File: web/api/v1/api.go [hunk 6]
```diff
@@ -1085,6 +1101,7 @@ func parseDuration(s string) (time.Duration, error) {
 
 func init() {
```
Comment (gouthamve) [inline web/api/v1/api.go]: Can you move the init to the top?

**Comment by gouthamve** [comment]
Hi @tariq1890, thanks for keeping at it, could you address @simonpasquier's comment about the `jsoniter` import and this is good to go!

**Comment by tariq1890** [comment]
Ah that keeps getting cleared lol due to auto-formatting by my IDE.

File: web/api/v1/api.go (hunk 6 above)
Comment (tariq1890) [reply web/api/v1/api.go]: Sure

**Comment by tariq1890** [comment]
@gouthamve Done :)

**Comment by gouthamve** [comment]
Thanks!

**Pull request merged**